### PR TITLE
Feature/group by conf

### PR DESCRIPTION
**Pull request by cmccully**
Adding support for named configuration modes. 

To deploy this into production, we should populate the previous images in the calimages table with the configuration_mode = '1x1' (or the appropriate binning).

Once this header keyword is ready to deploy into the field, we fill have to update the calimages table to have configuration_mode = 'full_frame'

### CODE REVIEW

File: banzai/images.py [hunk 1]
```diff
@@ -79,6 +79,10 @@ def __init__(self, runtime_context, filename=None, data=None, data_tables=None,
             self.gain = eval(str(header.get('GAIN')))
 
         self.ccdsum = header.get('CCDSUM')
+        self.configuration_mode = header.get('CONFMODE', file_utils.ccdsum_to_filename(self.ccdsum))
+        if self.configuration_mode == 'N/A':
```
Comment (drhaz) [inline banzai/images.py]: Make a comment to clarify that this is mitigating legacy?

File: banzai/mosaic.py [hunk 2]
```diff
@@ -18,13 +18,19 @@ def do_stage(self, image):
             nx, ny = get_mosaic_size(image, image.get_n_amps())
             mosaiced_data = np.zeros((ny, nx), dtype=np.float32)
             mosaiced_bpm = np.zeros((ny, nx), dtype=np.uint8)
+            x_detsec_limits, y_detsec_limits = get_detsec_limits(image, image.get_n_amps())
+            xmin = min(x_detsec_limits) - 1
+            ymin = min(y_detsec_limits) - 1
             for i in range(image.get_n_amps()):
+                ccdsum = image.extension_headers[i].get('CCDSUM', image.ccdsum)
```
Comment (drhaz) [inline banzai/mosaic.py]: It is understood that all extensions wil have the same binning. Do you want to assert this here?

File: banzai/mosaic.py [hunk 3]
```diff
@@ -40,6 +46,16 @@ def do_stage(self, image):
         return image
 
 
+def get_windowed_mosaic_slices(detsec, xmin, ymin, x_binning, y_binning):
+    unbinned_slice = fits_utils.parse_region_keyword(detsec)
```
Comment (drhaz) [inline banzai/mosaic.py]: Want to assert that  slice_width / binning == slice_width // binning ? That would be a safeguard against wrong detsec / binning config (although I would never do that)

File: banzai/utils/file_utils.py [hunk 4]
```diff
@@ -46,9 +46,20 @@ def instantly_public(proposal_id):
     return public_now
 
 
-def ccdsum_to_filename(image):
-    return 'bin{ccdsum}'.format(ccdsum=image.ccdsum.replace(' ', 'x'))
+def ccdsum_to_filename(ccdsum):
```
Comment (drhaz) [inline banzai/utils/file_utils.py]: Is there a particular reason to be inconsistent in the function parameters? It is an image object for filter_to_filename and config_to_filename, but the actual value for ccdsum_to_filename

File: banzai/images.py [hunk 5]
```diff
@@ -79,6 +79,10 @@ def __init__(self, runtime_context, filename=None, data=None, data_tables=None,
             self.gain = eval(str(header.get('GAIN')))
 
         self.ccdsum = header.get('CCDSUM')
+        self.configuration_mode = header.get('CONFMODE', file_utils.ccdsum_to_filename(self.ccdsum))
```
Comment (dmcollom) [inline banzai/images.py]: For clarity's sake, make the default value 'N/A'

File: banzai/mosaic.py [hunk 6]
```diff
@@ -48,16 +64,18 @@ def update_naxis_keywords(image, nx, ny):
     image.header['NAXIS2'] = ny
 
 
-def get_mosaic_size(image, n_amps):
+def get_detsec_limits(image, n_amps):
```
Comment (dmcollom) [inline banzai/mosaic.py]: Add docstring

File: banzai/tests/test_bias_comparer.py [hunk 7]
```diff
@@ -20,13 +20,14 @@ def test_null_input_image():
...
+    assert comparer.master_selection_criteria == ['configuration_mode']
 
 
 @mock.patch('banzai.calibrations.FRAME_CLASS', side_effect=FakeBiasImage)
 @mock.patch('banzai.calibrations.ApplyCalibration.get_calibration_filename')
-def test_returns_null_if_ccdsums_are_different(mock_cal, mock_frame):
-    handles_inhomogeneous_set(BiasComparer, FakeContext(), 'ccdsum', '1 1')
+def test_returns_null_if_configuration_modes_are_different(mock_cal, mock_frame):
+    handles_inhomogeneous_set(BiasComparer, FakeContext(), 'configuration_mode', 'central_2k_2x2')
+    handles_inhomogeneous_set(BiasComparer, FakeContext(), 'configuration_mode', 'central_2k_2x2')
```
Comment (dmcollom) [inline banzai/tests/test_bias_comparer.py]: Duplicate

File: banzai/utils/date_utils.py [hunk 8]
```diff
@@ -125,13 +125,14 @@ def get_dayobs(timezone):
 
 def validate_date(s):
     try:
-        return datetime.datetime.strptime(s, TIMESTAMP_FORMAT)
```
Comment (dmcollom) [inline banzai/utils/date_utils.py]: Merge issue, this should be changed back

File: banzai/images.py (hunk 1 above)
Comment (cmccully) [reply banzai/images.py]: Done.

File: banzai/utils/file_utils.py (hunk 4 above)
Comment (cmccully) [reply banzai/utils/file_utils.py]: It is called inside the image constructor so I had originally changed it, but doesn't need to be. Fixed.

File: banzai/images.py (hunk 5 above)
Comment (cmccully) [reply banzai/images.py]: Done.

File: banzai/utils/date_utils.py (hunk 8 above)
Comment (cmccully) [reply banzai/utils/date_utils.py]: Fixed.

File: banzai/tests/test_bias_comparer.py (hunk 7 above)
Comment (cmccully) [reply banzai/tests/test_bias_comparer.py]: Duplicate removed.

File: banzai/mosaic.py (hunk 2 above)
Comment (cmccully) [reply banzai/mosaic.py]: This should cover the case when they are different. For the moment, I think this makes sense.

File: banzai/mosaic.py (hunk 3 above)
Comment (cmccully) [reply banzai/mosaic.py]: Given the complexity of the indexing, I'm a little nervous to enforce that restriction. In this case, I'm inclined to just let the error get logged if a DETSEC keyword is incorrect rather than trying to code around every edge case.

File: banzai/mosaic.py (hunk 6 above)
Comment (cmccully) [reply banzai/mosaic.py]: Done.

File: banzai/tests/test_mosaic_maker.py [hunk 9]
```diff
@@ -9,6 +9,11 @@ class FakeMosaicImage(FakeImage):
     def __init__(self, *args, **kwargs):
         super(FakeMosaicImage, self).__init__(*args, **kwargs)
         self.extension_headers = None
```
Comment (mgdaily) [inline banzai/tests/test_mosaic_maker.py]: Still necessary to have `self.extension_headers` initialized to None here?

File: banzai/tests/test_dark_maker.py [hunk 10]
```diff
@@ -30,8 +30,8 @@ def test_header_cal_type_dark(mock_frame):
     assert images[0].header['OBSTYPE'].upper() == 'DARK'
 
 
-def test_returns_null_if_ccdsums_are_different():
-    handles_inhomogeneous_set(DarkMaker, FakeContext(), 'ccdsum', '1 1', calibration_maker=True)
+def test_returns_null_if_configuration_modes_are_different():
+    handles_inhomogeneous_set(DarkMaker, FakeContext(),'configuration_mode', 'central_2k_2x2', calibration_maker=True)
```
Comment (mgdaily) [inline banzai/tests/test_dark_maker.py]: Missing a space after the second comma.

Comment (cmccully) [reply banzai/tests/test_dark_maker.py]: Fixed.

File: banzai/tests/test_mosaic_maker.py (hunk 9 above)
Comment (cmccully) [reply banzai/tests/test_mosaic_maker.py]: Fixed.
